### DISCUSSION TITLE
Using submodules and serialization

### DISCUSSION THREAD

**Discussion by francis-switcho** [vuejs/pinia#2686] [Help and Questions] [answered]
Hi, I saw [this](https://github.com/vuejs/pinia/discussions/736) discussion here regarding store splitting and am currently struggling with serialization (JSON.stringify) due to this approach.

My use case is:
I have a store made by "assembling" smaller stores e.g. something similar to [this (playground)](https://play.pinia.vuejs.org/#eNqNVFFP2zAQ/itWNKmdaB0YE5NQiwoTk+ABpnWPeUmTS2tI7Mh2Qrco/31nO3HTrkW82f7uvvvubH9NcFuWtK4guA5mKpGs1ESBrkqSx3w9jwKtouAm4qwohdSkIZWCpRYSSEsyKQoyoqEye6rVaBcmIetx5EYg4ongCrlt7tzTjD/vEMn4mmUMUsSRYBwFUYBwxLOKJ5oJjsoki3P2t88lTcTJMJPWcV4Z/sfl8xPtgT9jJ/GT0rEGpGwN6yx0/WJ3uNFQlDmiuCNktqq0xnqLJGfJK07B5fOzMxzGA08kFMCxzcb1Qzlp21nokk4Q7ClHlmWvzVHsZ6esvmmaQV+G3hwa1QOlwSRwE58WcUlflOB4jXYmUQfg7V27KZmzknEWm5MoCO2agiqmKyneUB/mR8GkD13gvYUp1FqIXE3jkrm0jdalug7DJOUYn0LOakk56JCXRfhfzuKKXtGLMGerEAuFjKew3S+DGdMUio+w96GLc3rxlX6zrPVlR1oYVkOKV9viWLTCV5Wx9cFQElGULAf5XJr3tD+cOM/F26M907ICLzHZQPJ65PxFbZ3qnxJwfDXeqsd0LNegHXy/fIItrj1YiLTKMfod8BcokVdGowu7q3iKsgdxVu2DvWN8JL/V/VYDV31TRqidho23Y/7+Tus7uZf0cjDF/mfjAL0BpJAxfuAB9i0Nvr/5vh7sDQC2FnS/3dvIfEg4HhUx46MJwa89v3H6OncosGGQF511uJ13kIOoLy7qlgu9wagTwbyzmXN7JtH0JCdNX2jiySbE+EVrnag3qwMNh13Y1B9CnOwEMW9ymRCdz+1U7KJ8aV/4SFtH69/F8mR9xHz9VYzv6mh9E9XVD9p/AmseeA==)
I want to serialize and persist it (similar to what the [persisted state](https://github.com/prazdevs/pinia-plugin-persistedstate) plugin does)

Unfortunately I can't repro this on the playground, but I'll try and explain myself.

I am observing (on Vue 2.7.16, currently planning migration to 3) the same behaviour as described in #1149 , that is in dev mode no problem arises whereas in production` JSON.stringify` throws because of a circular structure.
This clearly happens because, even though I serialize the parent `store.$state`, the single members of this state are pointing to full store instances.

I believe [this](https://github.com/vuejs/pinia/blob/4a743fbf08cce13f84d9d29c59bf59b00ad7802e/packages/pinia/src/store.ts#L680) is what's preventing the error in dev mode (since the circular references appear to be under the `_p` property).

I managed to resolve this by omitting the `_p` property completely while serializing via the replacer function parameter in `JSON.stringify` but it doesn't feel like a solid approach.

Therefore I have some questions:
- Is this intended behavior, AKA is the way I'm "composing" the parent store wrong? E.g. Should I instead destruct inner stores (e.g. via `storeToRefs`) and expose only their state properties in the parent store?
- Is this somehow related to being on Vue 2 instead of Vue 3?

Thank you very much

Francis

**Comment by posva** [comment, marked as answer]
The issue in the playground is that the main store is returning the other stores. That creates state duplication and other other serialization issues. The solution is simple: do not return other stores (or their state) from a store.
